### PR TITLE
Add fine grained control over output buffers in the pipeline

### PR DESCRIPTION
**Pull request by JanuszL**
With current asynchronous executor design, we have n-buffers, while 1
is being consumed, other can be calculated. When n is equal to 2
and we release old one and start consuming new one. If released
one cannot be filled during one network forward and backward pass
everything waits until it is filled. By this path, it is possible to immediately release the buffer as soon as data is copied to some external
tensor. Also, this change adds the ability to set pipeline length in the executor
 - the default value is 2.

Signed-off-by: Janusz Lisiecki <jlisiecki@nvidia.com>

### CODE REVIEW

**Comment by JanuszL** [comment]
I need to update TF and examples as well.

File: dali/c_api/c_api.cc [hunk 1]
```diff
@@ -25,13 +25,15 @@ void daliCreatePipeline(daliPipelineHandle* pipe_handle,
     int length,
     int batch_size,
     int num_threads,
-    int device_id) {
+    int device_id,
+    int pipeline_length) {
```
Comment (ptrendx) [inline dali/c_api/c_api.cc]: That does not seem like a good name. prefetch_queue_depth perhaps?

File: dali/pipeline/executor/executor.cc [hunk 2]
```diff
@@ -238,9 +238,7 @@ void Executor::RunGPU() {
   previous_gpu_queue_idx_ = queue_idx;
 }
 
-void Executor::Outputs(DeviceWorkspace *ws) {
-  DALI_ENFORCE(ws != nullptr, "Workspace is nullptr");
-  ws->Clear();
+void Executor::ReleaseOutputs() {
```
Comment (ptrendx) [inline dali/pipeline/executor/executor.cc]: This is an API break - previously you could do runcpu/rungpu/outputs, now you need to do runcpu/rungpu/outputs/releaseoutputs.
At the very least you need to update all tests/benchmarks that will hang with this change. Is it not possible to have ReleaseOutputs as separate but also enable the old behavior?

File: dali/c_api/c_api.cc (hunk 1 above)
Comment (JanuszL) [reply dali/c_api/c_api.cc]: Done

File: dali/pipeline/executor/executor.cc (hunk 2 above)
Comment (JanuszL) [reply dali/pipeline/executor/executor.cc]: Added ShareOutputs and and implemented Outputs as call to ReleaseOutputs and ShareOutputs so old code should still work.

**Comment by JanuszL** [comment]
Builds after fixes 37408.

File: dali/python/nvidia/dali/pipeline.py [hunk 3]
```diff
@@ -230,35 +235,60 @@ def _run_gpu(self):
...
 
         If the pipeline was created with `exec_async` option set to `True`,
         this function will also start prefetching the next iteration for
         faster execution."""
+        if not self._built:
+            raise RuntimeError("Pipeline must be built first.")
+        if self._first_iter and self._exec_pipelined:
+            self._first_iter = False
+            for i in range(self._prefetch_queue_depth - 1):
+             self._start_run()
```
Comment (ptrendx) [inline dali/python/nvidia/dali/pipeline.py]: why not something like
```
             if self._first_iter:
                 _prefetch()
             else:
                 self._start_run()
```

File: dali/tensorflow/daliop.cc [hunk 4]
```diff
@@ -104,6 +104,11 @@ class DaliOp : public tf::OpKernel {
     OP_REQUIRES_OK(context, context->GetAttr("shape", &shape_));
     OP_REQUIRES_OK(context, context->GetAttr("num_threads", &num_threads));
     OP_REQUIRES_OK(context, context->GetAttr("device_id", &device_id));
+    if (context->HasAttr("prefetch_queue_depth")) {
```
Comment (ptrendx) [inline dali/tensorflow/daliop.cc]: You should add that attribute to REGISTER_OP (line 68)

**Pull request merged**